test/Libraries/Microsoft.Extensions.DependencyInjection.AutoActivation.Tests/AcceptanceTest.Keyed.cs

```diff
@@ -0,0 +1,425 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Collections.Generic;
+using System.Threading.Tasks;
+using Microsoft.Extensions.DependencyInjection.Test.Fakes;
+using Microsoft.Extensions.DependencyInjection.Test.Helpers;
+using Microsoft.Extensions.Hosting;
+using Microsoft.Extensions.Hosting.Testing;
+using Xunit;
+
+namespace Microsoft.Extensions.DependencyInjection.Test;
+
+public partial class AcceptanceTest
+{
+    [Fact]
+    public async Task CanAddAndActivateKeyedSingletonAsync()
+    {
+        var instanceCount = new InstanceCreatingCounter();
+        Assert.Equal(0, instanceCount.Counter);
+
+        var serviceKey = new object();
+        using var host = await FakeHost.CreateBuilder()
+            .ConfigureServices(services => services
+                .AddSingleton<IFakeServiceCounter>(instanceCount)
+                .AddActivatedKeyedSingleton<IFakeService, FakeService>(serviceKey))
+            .StartAsync();
+
+        var service = host.Services.GetKeyedService<IFakeService>(serviceKey);
```

Comment:
What if we also check that `instanceCount.Counter` equals to `1` before we get the service from the provider?